### PR TITLE
JBPM-5893: Some issues when using diagrams with no canvas root node.

### PR DESCRIPTION
**Pull request by romartin**
Hey @manstis , @hasys , @mbarkley 

Here are some bug fixes and a few refactors as well. 
See the [JIRA ticket](https://issues.jboss.org/browse/JBPM-5893).  

PS: The issues were discovered by @mbarkley - these issues were not present in BPMN neither CM, due to it's diagram settings, but those are blocker for Max as he cannot move forward.

Thanks!

### CODE REVIEW

**Comment by romartin** [comment]
@manstis BTW I had been doing a few refactors on BPMN/CM and the way they share the structure but using a different type of diagram bean. 
Created a `BPMNDiagram` interface with different implementations for CM/BPMN, but this interface can be now used on different places in order to share/reuse common logic for both definition sets. 
I think that some components, such as the oryx id mappings, and the graphs and diagram factories have been simplified this way.. lemme know if it's good for you! :)
Thanks!

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/graph/processing/traverse/content/ChildrenTraverseProcessorImpl.java [hunk 1]
```diff
@@ -72,68 +79,110 @@ protected boolean doStartEdgeTraversal(final Edge edge,
...
     @Override
     @SuppressWarnings("unchecked")
     protected boolean doStartNodeTraversal(final Node node,
                                            final ChildrenTraverseCallback<Node<View, Edge>, Edge<Child, Node>> callback) {
-        if (parents.isEmpty()) {
-            return super.doStartNodeTraversal(node,
-                                              callback);
+
+        if (!parentStack.isRootDefined() || parentStack.isRootPresent()) {
+            return firNodeTraverseCallback(node,
```
Comment (manstis) [inline kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/graph/processing/traverse/content/ChildrenTraverseProcessorImpl.java]: Typo "fir**e**Node..."

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/graph/processing/traverse/content/ChildrenTraverseProcessorImpl.java [hunk 2]
```diff
@@ -72,68 +79,110 @@ protected boolean doStartEdgeTraversal(final Edge edge,
...
+
+    @SuppressWarnings("unchecked")
+    private boolean firNodeTraverseCallback(final Node node,
+                                            final ChildrenTraverseCallback<Node<View, Edge>, Edge<Child, Node>> callback) {
+        if (!parentStack.isEmpty()) {
+            return callback.startNodeTraversal(parentStack.asList(),
                                                node);
+        } else {
+            callback.startNodeTraversal(node);
+            ;
```
Comment (manstis) [inline kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/graph/processing/traverse/content/ChildrenTraverseProcessorImpl.java]: Floating ```;```

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-project-api/src/main/java/org/kie/workbench/common/stunner/bpmn/project/factory/impl/BPMNProjectDiagramFactoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates.
```
Comment (manstis) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-project-api/src/main/java/org/kie/workbench/common/stunner/bpmn/project/factory/impl/BPMNProjectDiagramFactoryImpl.java]: 2017

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/definition/CaseManagementDiagram.java [hunk 4]
```diff
@@ -55,7 +55,7 @@
 // Unfortunately extending the foregoing and providing a new @CanContain annotation leads to problems with identifying
 // Factories for Definitions; as CM's BindableDefinitionAdapterProxy is then generated with support for the super-class.
 // This then leads the unmarshalling of model Elements to Definitions to use the wrong Factory and hence fail.
-public class BPMNDiagram implements BPMNDefinition {
+public class CaseManagementDiagram implements BPMNDiagram {
```
Comment (manstis) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/definition/CaseManagementDiagram.java]: Have you checked you can create, save and re-load CM diagrams?

The class ```org.kie.workbench.common.stunner.bpmn.backend.legacy.Bpmn20Stencil``` converts from our model classes to Oryx JSON stencils and uses the ```BPMNDiagram``` simple name to lookup an enum entry for our graph content. I have a feeling, renaming this class, could introduce issues.

Comment (manstis) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/definition/CaseManagementDiagram.java]: @romartin OK, tried locally and seems OK.. so just a few petty things to change :-)

Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/definition/CaseManagementDiagram.java]: yeahh I did a try to everything I think hehe, and it was working.. .:) BTW do you like that solution? or did you prefer the previous code?

Comment (manstis) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/definition/CaseManagementDiagram.java]: @romartin Oh, I'm happy. No problem :-)

Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/definition/CaseManagementDiagram.java]: good hehe :) lemme do that few updates and provide a new commit. Thanks!!

Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/definition/CaseManagementDiagram.java]: will cherry-pick to 7.0.x as well

**Comment by romartin** [comment]
Hey @manstis , added a new commit! Thanks again!

**Pull request merged**